src/mystery_event_script.c

```diff
@@ -0,0 +1,396 @@
+#include "global.h"
+#include "berry.h"
+#include "battle_tower.h"
+#include "easy_chat.h"
+#include "event_data.h"
+#include "mail.h"
+#include "mystery_event_script.h"
+#include "pokedex.h"
+#include "pokemon.h"
+#include "pokemon_size_record.h"
+#include "script.h"
+#include "constants/species.h"
+#include "strings.h"
+#include "string_util.h"
+#include "text.h"
+#include "util.h"
+#include "mystery_event_msg.h"
+#include "pokemon_storage_system.h"
+
+extern void sub_811EFC0(u8);
+extern void ValidateEReaderTrainer(void);
+
+extern ScrCmdFunc gMysteryEventScriptCmdTable[];
+extern ScrCmdFunc gMysteryEventScriptCmdTableEnd[];
+
+#define LANGUAGE_MASK 0x1
+#define VERSION_MASK 0x200
+
+EWRAM_DATA static struct ScriptContext sMysteryEventScriptContext = {0};
+
+static bool32 CheckCompatibility(u16 a1, u32 a2, u16 a3, u32 a4)
+{
+    if (!(a1 & LANGUAGE_MASK))
+        return FALSE;
+
+    if (!(a2 & LANGUAGE_MASK))
+        return FALSE;
+
+    if (!(a3 & 0x4))
+        return FALSE;
+
+    if (!(a4 & VERSION_MASK))
+        return FALSE;
+
+    return TRUE;
+}
+
+static void SetIncompatible(void)
+{
+    StringExpandPlaceholders(gStringVar4, gText_MysteryGiftCantBeUsed);
+    SetMysteryEventScriptStatus(3);
+}
+
+static void InitMysteryEventScript(struct ScriptContext *ctx, u8 *script)
+{
+    InitScriptContext(ctx, gMysteryEventScriptCmdTable, gMysteryEventScriptCmdTableEnd);
+    SetupBytecodeScript(ctx, script);
+    ctx->data[0] = (u32)script;
+    ctx->data[1] = 0;
+    ctx->data[2] = 0;
+    ctx->data[3] = 0;
+}
+
+static bool32 RunMysteryEventScriptCommand(struct ScriptContext *ctx)
+{
+    if (RunScriptCommand(ctx) && ctx->data[3])
+        return TRUE;
+    else
+        return FALSE;
+}
+
+void sub_8153870(u8 *script)
+{
+    InitMysteryEventScript(&sMysteryEventScriptContext, script);
+}
+
+bool32 sub_8153884(u32 *a0)
+{
+    bool32 ret = RunMysteryEventScriptCommand(&sMysteryEventScriptContext);
+    *a0 = sMysteryEventScriptContext.data[2];
+
+    return ret;
+}
+
+u32 RunMysteryEventScript(u8 *script)
+{
+    struct ScriptContext *ctx = &sMysteryEventScriptContext;
+    InitMysteryEventScript(ctx, script);
+    while (RunMysteryEventScriptCommand(ctx))
```

Comment:
Convention so far has been to keep the semicolon on the same line as the while statement, but visually separate the while statement from the surrounding lines.